### DISCUSSION TITLE
osu!catch editor ideas

### DISCUSSION THREAD

**Discussion by ekrctb** [ppy/osu#13534] [Ideas]
I'm opening this thread as a place to centralize ideas about osu!catch editor. Feel free to add ideas/suggestions.
I would especially appreciate small quality-of-life improvement things because such things are hard to notice for non-mappers.

Note: I don't have any implementation yet.

---

Edit note: the idea above is scraped. See <https://github.com/ppy/osu/discussions/13534#discussioncomment-985454>.

<details>
<summary>My idea (old)</summary>

My idea is there would be two "modes" in the compose screen.

### Classic

One is showing the hit objects in the 2D positions like osu!standard.
This is the preferred to edit the shape of sliders because it is very difficult to reason about the shape of the slider in the actual playfield.
In addition to the standard-like experience, catch specific things are:

- But irrelevant things like approach circles or combo numbers are not there. But probably some visualization of hit object ordering is desired?
- Only snap to the horizontal position.
- Showing hyper/non-hyper.

### Gameplay

It is showing the actual playfield/catcher in 1D position + 1D time.
The hit objects can be moved the obvious way.
Editing the shape of a slider (juice stream) is very hard to implement / under-specified, so the sliders are not editable in this mode at first.

- Editing a liner-type slider is feasible to implement.

Thoughts?
</details>

**Comment by peppy** [comment]
I'd heavily recommend keeping only one mode (the `Gameplay` mode in your proposal) and getting a minimum implementation working, with linear juice streams as you say. I think tackling the harder curve types may fall into place once we have all the other pieces working.

**Comment by Kimitakari** [comment]
Since osu!catch heavily relies on coordinates with x and y plus the distance snapping parameter I think it would be wise to implement those as well because us mappers/modders constantly rely on these coordinated where we place an object. Not to mention the ms snapping table which was created months ago in osu!catch Ranking Criteria would've been also implemented here (not even osu! has that function in lazer).

![](https://puu.sh/HPzeC/6304af5d9a.png)

**Reply by peppy** [reply]
These kinds of things would be represented graphically. Not based on pixel values and definitely not requiring a lookup table. Please get in the mindset of trying to make a user experience that is friendly to a new mapper, not trying to recreate the mess that we already have.

**Comment by Darius-Wattimena** [comment]
From my experience as catch mapper/modder/bn/nat I think the following QOL additions are essential to make me switch from the standard like editor to a catch specific one:
- Being able to show an indication for both hyperdashes and dashes, and display this on the fruit and juice stream. Knowing when something becomes a hyperdash or when something isn't catchable when walking would make mapping much more intuitive.
- Display hyperdash strength when clicking on the object (so trigger distance = 1.0 strength). Some rules and guidelines in the RC work with the strength of a hyperdash, so right now you have to calculate it yourself. This will make the mapper more aware of the strength of the hyperdashes they are using. (Doing this for walks and dashes would also be nice, but not sure how helpful it will be)
- Trigger distance button, basically a button that works like 'distance snap' but for hyperdashes, so you can easily create 1.0 strength or 1.5 strength hypers with ease.
- Being able to edit juice streams by their individual parts (e.g. slider tail, droplet).
- Being able to change the small droplet generation from random to actually following the path of the slider (I know that some mappers use a 3rd party program to make their droplets not random. Being able to do this in the client would be key if you want to make aesthetically pleasing juice streams)

**Reply by peppy** [reply]
- For the sake of people who aren't involved in community created lingo, could you please link to or document here what "strength" is? There's nothing in our code or documentation that mentions this.
- What does "edit juice streams by their individual parts" mean? Changing samples?

> * Being able to change the small droplet generation from random to actually following the path of the slider (I know that some mappers use a 3rd party program to make their droplets not random. Being able to do this in the client would be key if you want to make aesthetically pleasing juice streams)

This is out of scope and should also not be rankable. Jesus. I don't even want to know. This should be a standard feature request separate from the editor discussion and any beatmap using such techniques should probably not be ranked as there's no guarantee it won't break something. Please open a separate discussion if you want to discuss this further. It's probably something which would be considered in years, not now.

**Reply by Darius-Wattimena** [reply]
> * For the sake of people who aren't involved in community created lingo, could you please link to or document here what "strength" is? There's nothing in our code or documentation that mentions this.

Strenght here is referred to as `minimum distance needed for x movement * the desired strength`. For example, hyperdashes with a 1.0 strength are the trigger distance of a hyperdash, where trigger distance is the minimum distance needed to generate a hyperdash ([see RC glossary here](https://osu.ppy.sh/wiki/en/Ranking_Criteria/osu!catch#gameplay)). So let's say we have a 1.1 strength hyperdash you basically refer to `trigger distance * 1.1`.

The strength of a dash on the other hand is indicated by the minimum distance when the dash key is needed to be pressed also referred to as the 'minimum dash distance'. 

The strength of a walk is indicated by the ideal walk speed of the catcher. From what I remember to get that you do `300 / bpm` to find the ideal SV and then having a 1.0x distance will be your ideal walk.

> * What does "edit juice streams by their individual parts" mean? Changing samples?

Here I refer to being able to edit the position of a slider head, big droplet or slider end/repeat without having to change the slider velocity and/or slider velocity multiplier. It might also be out of scope as this will actually change the way how sliders are generated mid slider path (if possible at all)?

> This is out of scope and should also not be rankable. Jesus. I don't even want to know. This should be a standard feature request separate from the editor discussion and any beatmap using such techniques should probably not be ranked as there's no guarantee it won't break something. Please open a separate discussion if you want to discuss this further. It's probably something which would be considered in years, not now.

Good point, I actually wrote everything that came to my mind. About it not being rankable is actually quite a concern as some users are using this tool, [find the github of the project here](https://github.com/Phob144/DropletDerandomizer). But if it should be unrankable then I'll take action on this, and update the RC for it 👍

**Reply by peppy** [reply]
> Here I refer to being able to edit the position of a slider head, big droplet or slider end/repeat without having to change the slider velocity and/or slider velocity multiplier. It might also be out of scope as this will actually change the way how sliders are generated mid slider path (if possible at all)?

We do plan (even for osu!, see WIP at https://github.com/ppy/osu/pull/10417) to make it easier to adjust the slider velocity per hit object without creating timing points. Is that what you're asking for, as opposed to new functionality allowing changing juice streams in a way you could until now?

**Reply by Darius-Wattimena** [reply]
> We do plan (even for osu!, see WIP at #10417) to make it easier to adjust the slider velocity per hit object without creating timing points. Is that what you're asking for, as opposed to new functionality allowing changing juice streams in a way you could until now?

This is not 100% the same thing. What I'm asking is to manipulate the slidervelocity of a slider mid sliderpath by for example simply dragging any of the slider parts (e.g. slider start, slider end, big droplet).

**Reply by peppy** [reply]
That sounds like another new feature and format change. That's not what we're looking to do here. We're looking to make an editor with feature parity. Please refrain from asking for new game mechanics in this thread.

**Comment by ekrctb** [comment]
After some thinking about juice stream editing, only allowing polygonal chains (only using the `Linear` control point type) should be the way.
Reasonings are:

1. Any juice stream pattern can be represented by a polygonal chain.
    - It typically only requires no more than four vertices to represent a pattern generated by a curved slider path by a polygonal chain.
2. Because the smooth curve of the slider is not visible, but only sampled droplet positions affect gameplay, it doesn't matter if the same juice stream is represented by one circular path or a rectangle (for example).
    - Critically, that implies the smooth path of the slider shouldn't be visible in the lazer editor. That is, the while line added in <https://github.com/ppy/osu/pull/13795> should be removed.
    - Instead, the slider path should be visualized as a sequence of points sampled at droplet generating times (perhaps including tiny droplets).
3. Even if there might be a more appropriate representation of juice stream patterns in the lazer editor, the result must be exportable to a legacy .osu format. Furthermore, in the current lazer editor, the edit result is serialized/deserialized via the legacy beatmap format.
    - I considered adding metadata to the .osu format in a backward-compatible manner, for example by "magic comments". However, such metadata will be lost when the map is edited in the stable editor (for example, consider when a modder using the stable editor). I consider it unacceptable, at least for now.
    - Therefore, the juice stream shape information must resist a round-trip to/from the .osu legacy format.
4. Uses of complex curved sliders in osu!catch mapping is less common than simple linear ones.
    - Curved paths are not natural for the movement because they require changing the speed of the catcher while catching the juice stream (tapping the dash key or arrow keys).
    - If required, shortcuts to commonly used patterns, like the shape of parenthesis `(` or `)` can be added.

**Reply by peppy** [reply]
I think something like this sounds fine for now, until we figure out the path forward with beatmap format changes (probably 6+ months away).
